Unresolved symbols assertion during pickling when using a 'by name' expression that defines a trait inside

In my attempts to port a macro to Scala 3, I stumbled on an issue that seems very similar to #5572 which was reported and fixed by @milessabin. Using his repro sample, I was able to reproduce the bug on 3.0.0-M1 ([try it in Scastie](https://scastie.scala-lang.org/tdZR1JvlSvK9q0z1DOAFvw))

When defining a trait inside the block passed to `foo`, if used with an `Array`, it will blow up with the exception below.
Changing the array to, say, a `List` or a `Vector` does not cause the issue, and moving the trait definition outside of the block also does not cause the issue.

Also seems that changing the by-name parameter to a lambda (e.g. `() => A`) does not cause the issue either.

The original code crashed when defining a Vector of those trait instances, followed by a call to `.toArray`, which caused the crash. I suspect this has possibly something to do with a `ClassTag`?

## Minimized code

```Scala
trait Foo {
  inline def foo[A](t: => A): Unit = ()
}

object Bar extends Foo

object Test {
  Bar.foo {
    trait T1
    val array = Array(new T1 {})
  }
}
```

## Output (click arrow to expand)
<details>

Scastie output:
```
java.lang.AssertionError: assertion failed: unresolved symbols: [33mtrait[0m [35mT1[0m(line 8) when pickling /tmp/scastie7347053754411828221/src/main/scala/main.scala
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:765)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$3$$anonfun$2(Pickler.scala:69)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:105)
        ...
```
</details>

--- a/compiler/src/dotty/tools/dotc/ast/TreeTypeMap.scala
+++ b/compiler/src/dotty/tools/dotc/ast/TreeTypeMap.scala
@@ -4,7 +4,7 @@ package ast
 
 import core._
 import Types._, Contexts._
-import Symbols._, Annotations._, Trees._, Symbols._
+import Symbols._, Annotations._, Trees._, Symbols._, Constants.Constant
 import Decorators._
 import dotty.tools.dotc.transform.SymUtils._
 import core.tasty.TreePickler.Hole
@@ -124,6 +124,8 @@ class TreeTypeMap(
           cpy.Labeled(labeled)(bind1, expr1)
         case Hole(isTermHole, n, args) =>
           Hole(isTermHole, n, args.mapConserve(transform)).withSpan(tree.span).withType(mapType(tree.tpe))
+        case lit @ Literal(Constant(tpe: Type)) =>
+          cpy.Literal(lit)(Constant(mapType(tpe)))
         case tree1 =>
           super.transform(tree1)
       }

--- a/compiler/src/dotty/tools/dotc/core/Symbols.scala
+++ b/compiler/src/dotty/tools/dotc/core/Symbols.scala
@@ -835,8 +835,10 @@ object Symbols {
           info = completer,
           privateWithin = ttmap1.mapOwner(odenot.privateWithin), // since this refers to outer symbols, need not include copies (from->to) in ownermap here.
           annotations = odenot.annotations)
-        copy.registeredCompanion =
-          copy.registeredCompanion.subst(originals, copies)
+        copy.denot match
+          case cd: ClassDenotation =>
+            cd.registeredCompanion = cd.unforcedRegisteredCompanion.subst(originals, copies)
+          case _ =>
       }
 
       copies.foreach(_.ensureCompleted()) // avoid memory leak

--- a/tests/pos/i10542.scala
+++ b/tests/pos/i10542.scala
@@ -1,0 +1,29 @@
+package test1:
+
+  trait Foo {
+    inline def foo[A](t: => A): Unit = ()
+  }
+
+  object Bar extends Foo
+
+  object Test {
+    Bar.foo {
+      sealed trait T1
+      case object S1 extends T1
+    }
+  }
+/*
+package test2:
+
+  trait Foo {
+    inline def foo[A](t: => A): Unit = ()
+  }
+
+  object Bar extends Foo
+
+  object Test {
+    Bar.foo {
+      trait T1
+      val array = Array(new T1 {})
+    }
+  }*/

--- a/tests/pos/i10542.scala
+++ b/tests/pos/i10542.scala
@@ -1,4 +1,4 @@
-package test1:
+package test_10542:
 
   trait Foo {
     inline def foo[A](t: => A): Unit = ()
@@ -12,8 +12,8 @@ package test1:
       case object S1 extends T1
     }
   }
-/*
-package test2:
+
+package test_10540:
 
   trait Foo {
     inline def foo[A](t: => A): Unit = ()
@@ -26,4 +26,13 @@ package test2:
       trait T1
       val array = Array(new T1 {})
     }
-  }*/
+  }
+
+package test_9655:
+
+  inline def foo[T](inline body: T): T = body
+
+  def test = foo {
+    sealed trait Status
+    object Active extends Status
+  }

